### ISSUE TITLE
GSDK re-architecture?

### ISSUE THREAD

**Issue by dgkanatsios** [PlayFab/gsdk#54]
As a follow-up to the comments [here](https://github.com/PlayFab/gsdk/pull/52#issuecomment-625516040) and [here](https://github.com/PlayFab/gsdk/pull/52#issuecomment-625517542) I'm opening this issue to discuss potential re-architecture of the GSDK. 
@darvell thanks for bringing this up, do you want to add what we could do better? Also cc  @MScottAdams @CaptainPineapple

**Comment by darvell** [comment]
I think before we discuss this, scope needs to be brought in. In my opinion there is functionality on the backend that I would need to know would be considered before throwing ideas out there.

A simple example could be moving communications between the AgentVM and the game server itself to a websocket. Maybe even throw gRPC on top of that so I, a developer on the dashboard, could administrate my servers/matches right there rather than building this monolith secondary layer that tries to smuggle data through PlayFab via the heartbeat metadata or an external service.

If we're talking just purely the GSDK with no implementation changes per se but rather just pure code architecture, well, then we can start from there.

**Comment by darvell** [comment]
First point of contention, the Unity SDK and the normal C# SDK both in PlayFab as a whole and GSDK are completely different. One offers async and is more up to date as I believe it's built from the xplat stuff and the unity isn't.

We have async in Unity. I can tell you that unity developers are on .NET 4.5 and writing some awful wrappers to turn all these calls in to tasks. We should be aiming for the removal of anything Unity specific in the GSDK since having this magic Unity only version doesn't actually provide anything of use and adds technical debt.

We don't need monobehaviors, all of these are HTTP call wrappers or static monobehaviors just "chilling" in their own space under the guise of a game object, al alone essentially 'marshalling' over events in to Unity land which is over the top.

I'll make this point blunt. I don't think we can improve the GSDK nor the PlayFab SDK without having to say "this is the last time we make a super unity specific version" and move to the standard CSharp library as the one. If for some reason we need to add support for some defines when in Unity, okay, so be it. I'd rather have the one C# SDK to manage that generates a `#if UNITY` for any small code blocks.

Even then, we have our older users on .NET 3.5 and no tasks. Can we not just auto generate for them methods that go the callback way as second class citizens?

If the standard PF SDK can get reduced from being both Unity and C# to just C# that happens to work in Unity, we can potentially move the GSDK in to there as well and either enable/diable it as we please. Maybe as a submodule.

Other platforms have it good but after a while in production dealing with GSDK/PF SDK, the Unity integration is confusing and doesn't conform to the normal C# SDK which I find myself using for my web backend project.

My initial proposal would be, with the normal C# GSDK that's auto-generated, we should look _why_ it doesn't work and what's required. .NET 3.5 is marked as deprecated in Unity 2018.3 onwards so, let's not fear TPL and remove callback hell.

**Comment by darvell** [comment]
Adding on to the prior note of 'touch Unity less', we just accepted a PR because along the way, a basic polling method decided it needs be a monobehaviour forgetting that we as gamedevs, will change scenes and can't lose that data. From personal anecdote, I had to build a lot of backend services to bypass getting some information from the heartbeat because of a lack of consistency in data and it changing more often than it should and that was due to scene loading. Yes, that may be fixed now but our option was really just making it static when we don't even need to be touching Unity (outside of maybe a background job/thread)?

Simply put and maybe my overall thing I'll be arguing for is that I don't believe the GSDK should even be split out. It's part of the PF SDK, it leverages a lot in there, why can't we hide it behind a define? You're running a thread and polling endpoints. There's nothing that makes the GSDK exactly unique and detached.

**Comment by darvell** [comment]
Final for now, as discussed in a phone call a couple months ago, there are some developers out there who aren't using the GSDK as intended, those who don't intend on lobbies or the like but rather using Thunderhead (sorry, I'm sticking to the codename always) as a pure _hosting_ platform. While under NDA and it's not my code to show, the amount of effort behind it is why I can say other bigger indie devs just don't bother and lift whatever is in Steamworks instead. There's issues in price   as well but that's out of scope.

This is best spoken about with real world cases which after having to be out of gamedev for some time due to unfortunate circumstances, is a bit difficult. But, another goal of architecture should be to reduce the Xbox Live influence of how it's all P2P (but with a secret server behind it) and give the GSDK an easier way to go "I can hold 32 players and have 32 tickets. Take one if you want one and join." Having to do mandatory matchmaking rather than just getting a list of possible servers (via PlayFab) and simply asking if it'd be okay to join the server (it's not full or it hasn't reported that it's not taking players) would be a great addition.

In our project Soviet Jump Game, we worked around a lot of this and rather than having to queue up, build a lobby, etc, we can just say "here's the server that's best for you, here's a ticket, go join". Backfill tickets don't seem to work as intended as it runs off the idea that Thunderhead servers are only for people who use the matchmaking which really limits use cases.

I do question what is the appropriate approach for all of this. Does the GSDK need changing or does there need to be a community abstraction layer on top that allows for flexibility in how the system is used? Even with abstractions though, that doesn't account for failings in the backend like being able to modify say, server metadata for a single instance on the fly without building a third party service.

Maybe an open source .NET Core server with a drop in client library if GSDK is added to the game project to allow developers to have more dynamic information that can update in real time is something that also needs to be done.

Lots of rambling, heavy mind.  There's just definitely room for adding flexibility and making this all much nicer. Maybe we could arrange code review sometime soon and have some kind of zoom meeting with any other interest developers from other studios willing to talk about problems on their end.
